crates/holochain_p2p/src/op_store.rs

```diff
@@ -0,0 +1,408 @@
+                            let mut rows = match stmt.query(named_params! {
+                                ":storage_start_loc": arc_start,
+                                ":storage_end_loc": arc_end,
+                                ":timestamp_min": latest_timestamp.as_micros(),
+                                // Fetch ops in batches of 500. This lets us observe the `limit_bytes`
+                                // without going to the database too many times.
+                                // Because the timestamp being queried is the integration timestamp,
+                                // it shouldn't be possible for >500 ops authored at the same time
+                                // to prevent this loop from proceeding.
+                                ":limit": 500,
+                            }) {
+                                Ok(rows) => rows,
+                                Err(e) => return Err(e.into()),
+                            };
+
+                            let ops_size = out.len();
+
+                            while let Some(row) = rows.next()? {
+                                let hash: DhtOpHash = row.get(0)?;
+                                let timestamp = Timestamp::from_micros(row.get::<_, i64>(1)?);
+                                let serialized_size: u32 = row.get(2)?;
+
+                                if used_bytes + serialized_size > limit_bytes {
+                                    break 'outer;
+                                }
+
+                                let op_id = OpId::from(Bytes::copy_from_slice(hash.get_raw_36()));
+                                if out.insert(op_id) {
+                                    latest_timestamp = timestamp;
+                                    used_bytes += serialized_size;
+                                }
+                            }
+
+                            // If we didn't discover any new ops, break
+                            if out.len() == ops_size {
+                                break;
+                            }
+                        }
+
+                        Ok((out.into_iter().collect(), used_bytes, latest_timestamp))
+                    },
+                )
+                .await
+                .map_err(|e| K2Error::other_src("Failed to retrieve op ids bounded", e))?;
+
+            Ok(out)
+        })
+    }
+
+    fn store_slice_hash(
+        &self,
+        arc: DhtArc,
+        slice_index: u64,
+        slice_hash: Bytes,
+    ) -> BoxFuture<'_, K2Result<()>> {
+        let db = self.db.clone();
+
+        let (arc_start, arc_end) = match arc {
+            DhtArc::Empty => {
+                return Box::pin(async move { Ok(()) });
+            }
+            DhtArc::Arc(start, end) => (start, end),
+        };
+
+        Box::pin(async move {
+            db.write_async(move |txn| -> StateMutationResult<()> {
+                let mut stmt = txn.prepare(
+                    r#"INSERT INTO SliceHash
```

Comment:
Why are these statement written inline?

Reply:
They seemed small enough to have in strings and not very re-usable. I can move if you'd prefer consistency here?

Reply:
No, all good.